kube-client/src/discovery/apigroup.rs

```diff
@@ -327,7 +361,145 @@ impl ApiGroup {
+                ],
+                freshness: Some("Current".to_string()),
+            }],
+        };
+
+        let group = ApiGroup::from_v2(&ag).unwrap();
+
+        assert_eq!(group.name(), "apps");
+        assert_eq!(group.preferred_version(), Some("v1"));
+        assert_eq!(group.versions().collect::<Vec<_>>(), vec!["v1"]);
+
+        let resources = group.recommended_resources();
+        assert_eq!(resources.len(), 2);
+
+        let (deploy_ar, deploy_caps) = group.recommended_kind("Deployment").unwrap();
+        assert_eq!(deploy_ar.group, "apps");
+        assert_eq!(deploy_ar.version, "v1");
+        assert_eq!(deploy_ar.api_version, "apps/v1");
+        assert_eq!(deploy_ar.kind, "Deployment");
+        assert_eq!(deploy_caps.scope, Scope::Namespaced);
+    }
+
+    #[test]
+    fn test_api_group_from_v2_core() {
+        let ag = APIGroupDiscovery {
+            metadata: Some(ObjectMeta {
+                name: Some("".to_string()), // core group has empty name
+                ..Default::default()
+            }),
+            versions: vec![APIVersionDiscovery {
+                version: Some("v1".to_string()),
+                resources: vec![
+                    make_v2_resource("pods", "Pod", "Namespaced", vec!["get", "list", "watch"]),
+                    make_v2_resource("nodes", "Node", "Cluster", vec!["get", "list"]),
+                ],
+                freshness: Some("Current".to_string()),
+            }],
+        };
+
+        let group = ApiGroup::from_v2(&ag).unwrap();
+
+        assert_eq!(group.name(), "");
+        assert_eq!(group.preferred_version(), Some("v1"));
+
+        let (pod_ar, pod_caps) = group.recommended_kind("Pod").unwrap();
+        assert_eq!(pod_ar.group, "");
+        assert_eq!(pod_ar.api_version, "v1"); // core group: no prefix
+        assert_eq!(pod_caps.scope, Scope::Namespaced);
+
+        let (node_ar, node_caps) = group.recommended_kind("Node").unwrap();
+        assert_eq!(node_ar.kind, "Node");
+        assert_eq!(node_caps.scope, Scope::Cluster);
+    }
+
+    #[test]
+    fn test_api_group_from_v2_multiple_versions() {
+        let ag = APIGroupDiscovery {
+            metadata: Some(ObjectMeta {
+                name: Some("batch".to_string()),
+                ..Default::default()
+            }),
+            versions: vec![
+                // First version is preferred
+                APIVersionDiscovery {
+                    version: Some("v1".to_string()),
+                    resources: vec![make_v2_resource("jobs", "Job", "Namespaced", vec!["get", "list"])],
+                    freshness: Some("Current".to_string()),
+                },
+                APIVersionDiscovery {
+                    version: Some("v1beta1".to_string()),
+                    resources: vec![make_v2_resource("jobs", "Job", "Namespaced", vec!["get"])],
+                    freshness: Some("Current".to_string()),
+                },
```

Comment:
nit: this test might break on us in the future it because it is verifying a property about beta versions. [they removed this version for cronjob](https://kubernetes.io/docs/reference/using-api/deprecation-guide/#v1-25) earlier, so this one is not safe.

on the other hand major versions are meant to [never be removed[(https://kubernetes.io/docs/reference/using-api/deprecation-policy/) (as per `Rule #4a`). So maybe it's better to check the `autoscaling` group which has 2 separate major versions?